### DISCUSSION TITLE
Harmony batch integration with Spectre?

### DISCUSSION THREAD

**Discussion by ebro17** [ImmuneDynamics/Spectre#115] [Help]
Hi, 

I'm wondering if there is a way to use Harmony rather than CytoNorm for the batch correction step in the discovery workflow with batch alignment? I know this is typically used for RNA-seq data, however it seems like it should work well with CyTOF data and I am relatively new to R so it's not clear to me if there's an easy way to implement this into the Spectre workflow.  

Thanks in advance!
- E

**Comment by tomashhurst** [comment]
Hey @ebro17 , great question!

We actually have Harmony in Spectre already, so you are in luck! I don't have a workflow written up for it, but I have been using it in the majority of my analysis recently. The function is `run.harmony()` -- you can check the argument descriptions using `?? run.harmony` to figure out the inputs required.

Three important things to keep in mind:

1. Harmony doesn't adjust cellular expression levels, but pulls the data into what is effectively a shared 'PCA' space. As a result, once the integration is done, you will need to run clustering/UMAP etc using those columns.

2. Because Harmony doesn't use any controls, how it integrates data is very independent. However, in my experience, it has been very good with cytometry data. Nevertheless, it is important to carefully assess the output.

3. One argument in Harmony is called 'npcs' which refers to the number of principle components it will use. I _think_ the default is 30, which means if you are using less than 30 columns you would need to adjust this to n-1 (i.e. if you were using 10 columns, you would need to specify `npcs = 9`.

Good luck! Let us know if you need any help getting it working.

Tom

**Reply by ebro17** [reply]
This is great, thank you! I'm wondering if you know what could be causing the following error? "Warning: Quick-TRANSfer stage steps exceeded maximum (= 16798500)". It doesn't actually stop running with an error, but I've let it sit and run for a while and it never finishes. I'm using the following: 

npcs <- length(cellular.cols)/2 - 1
harmony.dat <- run.harmony(samples.dat, align.cols = cellular.cols, batch.col = "Batch", npcs = npcs, max.iter.harmony = 100, max.iter.cluster = 100, plot_convergence = TRUE)

Thank you again for the help.

**Comment by tomashhurst** [comment]
@ebro17 how many cells are you running? Harmony, despite how well it works, will still take some time to run with >10^5 cells. If you are running millions of cells it will take quite some time.

**Comment by jamila-griffith** [comment]
Can we use Harmony for the high content microscopy data? I wonder if someone has tried that before.

**Comment by tomashhurst** [comment]
Hi @jamila-griffith , 

We've certainly tried Harmony and other integration methods on segmented imaging data (i.e. on the cellular data derived from images after cell segmentation). Variable success in my hands, as the degree of cellular overlap, and how much this might change in different images, leads to some pretty significant challenges in modelling and correcting for differences across batches. As an example, perhaps in image 1 there are T and B cells, but they don't really overlap with each other, so CD3 and CD19 are essentially independent. However, in an area of dense infiltrate there will be cells apparently expressing both due to the spatial overlap, and this is difficult for such an algorithm to deal with.

**Reply by jamila-griffith** [reply]
Is there any protocol available and guide for how to apply this on the cellular data?

**Reply by jamila-griffith** [reply]
I am trying to reduce the batch effect from quantitative features derived from microscopy images and am wondering if the algorithm would still work for that. In the quantitative data the overlap does not matter because I extract features from single cell e.g. cell area, size, shape, morphology, intensity etc. This is one of the big challenge of the images that we are facing. Would be good other people's experience and what methods if any have they used.
